docs/design/features/MarshalDirective.md

```diff
@@ -0,0 +1,206 @@
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public sealed class MarshalDirectiveAttribute : Attribute
+    {
+        /// <summary>
+        /// Default constructor.
+        /// </summary>
+        public MarshalDirectiveAttribute()
+        {
+            this.Type = MarshalDirectiveType.Default;
+        }
+
+        /// <summary>
+        /// The directive type.
+        /// </summary>
+        /// <remarks>
+        /// The default value is <see cref="MarshalDirectiveType.Default "/>.
+        /// </remarks>
+        public MarshalDirectiveType Type { get; set; }
+
+        /// <summary>
+        /// The type to use to search for user defined functions.
+        /// </summary>
+        public Type UserFunctionType { get; set; }
+
+        /// <summary>
+        /// The name of the user defined stub to call.
+        /// Only used if <see cref="MarshalDirectiveAttribute.Type"/> is defined as
+        /// <see cref="MarshalDirectiveType.UserDefinedStub"/>.
+        /// </summary>
+        /// <remarks>
+        /// The supplied name will be searched for on
+        /// <see cref="MarshalDirectiveAttribute.UserFunctionType"/>.
+        ///
+        /// The supplied function must conform to a signature of that matches
+        /// the managed definition of the function, except for the first argument
+        /// which will be supplied by the runtime. The first argument will be
+        /// a resolved native function pointer.
+        ///
+        /// For example, given the following P/Invoke signature:
+        /// <code>
+        /// [DllImport("Kernel32.dll", ExactSpelling = true)]
+        /// [MarshalDirectiveAttribute(
+        ///         Type = MarshalDirectiveType.UserDefinedStub, 
+        ///         UserFunctionType = typeof(Stubs),
+        ///         UserDefinedStub = nameof(QueryPerformanceCounterStub))]
+        /// extern static unsafe bool QueryPerformanceCounter(long* lpPerformanceCount);
+        /// </code>
+        ///
+        /// The supplied user defined stub should be:
+        /// <code>
+        /// static unsafe bool QueryPerformanceCounterStub(
+        ///     IntPtr nativeFunctionPtr,
+        ///     long* lpPerformanceCount) { ... }
+        /// </code>
+        ///
+        /// During the call to the stub, the supplied <see cref="System.IntPtr"/>
+        /// would be a native function pointer.
+        /// </remarks>
+        public string UserDefinedStub { get; set; }
+    }
+}
+```
+
+## Questions
+
+* Can the above API be used to provide a reverse P/Invoke stub?
+* Does the the Marshal Directive have any interaction with the [`SuppressGCTransitionAttribute`](https://github.com/dotnet/runtime/issues/30741)?
+* Should the `MarshalDirectiveAttribute` be considered in potential support for the [typing of function pointers](https://github.com/dotnet/roslyn/issues/39865#issuecomment-600884703)?
```

Comment:
Attributes cannot be used to typing of function pointers as we have discussed.

Reply:
Agreed on attributes, but that is the same statement for `SuppressGCTransitionAttribute`. My point here was we have discussed the concept of `SuppressGCTransitionAttribute`  being applied to function pointers so does this align with that as well or not.

Reply:
This one has arguments. `SuppressGCTransitionAttribute` is just a flag without arguments. The modopt approach that we have discussed works for flags. It does not work for anything with arguments - the encoding does not allow it.

Reply:
Yep. Agreed in its current form there are problems. The concept is my question. Does providing function pointers with a mechanism to define a specific marshal stub make sense. In this case the proposal is for such a mechanism and does function pointers warrant a discussion about that or is that entire scenario different enough that there isn't any need to concern ourselves with including it in this kind of proposal.